### ISSUE TITLE
Bankster: `claim deposit` operation changes

### ISSUE THREAD

**Issue by rwrzesien** [golemfactory/concent#895]
There should be some updates in the current implementation of `claim deposit` operation in Bankster to allow it to keep deposit state.

##### Input - modify current input by adding following parameters
| Parameter                       | Type                  | Optional | Remarks                                                                                                                 |
|---------------------------------|-----------------------|----------|-------------------------------------------------------------------------------------------------------------------------|
| `subtask`                       | `Subtask`             | no       | Subtask object.
| `requestor_public_key`          | string                | no       | Public key of the requestor. Comes from `TaskToCompute`.
| `provider_public_key`           | string                | no       | Public key of the provider. Comes from `TaskToCompute`.
| `requestor_ethereum_public_key` | string                | no       | Address of the Ethereum account belonging to the requestor. Cannot be the same as `provider_ethereum_public_key`. Comes from `TaskToCompute`.
| `provider_ethereum_public_key`  | string                | no       | Address of the Ethereum account belonging to the provider. Cannot be the same as `requestor_ethereum_public_key`. Comes from `TaskToCompute`.
| `subtask_cost`                  | decimal               | no       | The cost of performing the work on the subtask that the requestor has agreed to pay the provider. Comes from `TaskToCompute.price` multiplied by maximum task duration. Must be greater than zero.

##### Output - replace current output with:
| Parameter                       | Type                  | Optional | Remarks                                                                                                                 |
|---------------------------------|-----------------------|----------|-------------------------------------------------------------------------------------------------------------------------|
| `claim_against_requestor`       | `DepositClaim`        | no       | The `DepositClaim` object that was created to lock a part of requestor's deposit.
| `claim_against_provider`        | `DepositClaim`        | yes      | The `DepositClaim` object that was created to lock a part of provider's deposit. May be empty if nothing is claimed from the provider.

##### Sequence of operation - modify current sequence:
1. **Claim calculation**: Bankster determines the amount that needs to be claimed from each account (This should be already done, but please review and confirm):
    - In `ForcedAcceptance` use case:
        - Requestor may have to pay `subtask_cost`.
        - Provider does not pay anything.
    - In `AdditionalVerification` use case:
        - Requestor may have to pay `subtask_cost`.
        - Provider needs to pay verification cost (which is constant and determined by Concent's settings).
2. **Initialization**:
    - Bankster begins a database transaction.
    - Bankster creates `Client` and `DepositAccount` objects (if they don't exist yet) for the requestor and also for the provider if there's a non-zero claim against his account.
    - Bankster commits the database transaction to prevent these objects from being rolled back in case of failure.
3. **Deposit query** (This should be already done, but please review and confirm):
    - Bankster asks SCI about the amount of funds available in requestor's deposit.
    - If the amount claimed from provider's deposit is non-zero, Bankster asks SCI about the amount of funds available in his deposit.
    - The check is intentionally performed outside of a database transaction.
        It does not affect the database and it's a relatively slow operation (requires one or more HTTP requests).
4. **Claim freeze**:
    - Bankster begins a database transaction.
    - Bankster puts database locks on all `DepositAccount` objects that will be used as payers in newly created `DatabaseClaim`s.
5. **Requestor's spare deposit check** (This is partially done, so please apply only necessary updates):
    - Bankster sums the `amount`s of all existing `DepositClaim`s where the requestor is the payer.
    - **Claims against requestor's deposit can be paid partially** because the service has already been performed by the provider and giving him something is better than giving nothing.
        If the existing claims against requestor's deposit are greater or equal to his current deposit, we can't add a new claim.
        - Bankster responds with `claim_against_requestor` and `claim_against_requestor` being both `None`.
    - Otherwise Bankster continues.
6. **Provider's spare deposit check**: If the amount claimed from his deposit is non-zero (This is partially done, so please apply only necessary updates).
    - Bankster sums the `amount`s of all existing `DepositClaim`s where the provider is the payer.
    - **Claims against provider's deposit must be paid in full** because they're payments for using Concent and we did not perform the service yet so we can just refuse.
      If the total of existing claims **and** the current claim is greater or equal to the current deposit, we can't add a new claim.
        - Bankster removes the `DepositClaim` just created aginst requestor's deposit.
        - Bankster responds with `claim_against_requestor` and `claim_against_requestor` being both `None`.
    - Otherwise Bankster continues.
7. **Deposit lock**: For both provider and requestor, if the amount claimed from the deposit is non-zero:
    - Bankster creates a `DepositClaim` object
        - `payee_ethereum_public_key`: payee can be either the provider (if the requestor is the payer) or Concent (if the provider is the payer).
            The address of Concent's account either comes from settings or is automatically inserted by the contract.
        - `amount`: The full amount claimed.
            Even if there's not enough deposit, we don't reduce the claim until the moment when we actually create a transaction.
            The situation may change in the meantime - some claims may be canceled or the client may increase (or reduce) the deposit.
        - `tx_hash`: We're not creating a transaction yet so this field is left empty.
8. **Unfreeze and result**: If everything goes well, the operation returns (Replace currently returned results).
    - All the created `DepositClaim`s are included in the result.
        At least one claim must have been created.
    - Bankster commits the database transaction.
        Database locks on `DepositAccount`s are released.

##### Expected Concent Core behavior - make sure that:
- Concent Core is responsible for removing `DepositClaim` objects if it fails at any point after the operation if the operation has succeeded.
    - Bankster removes `DepositClaim` objects only after they're successfully paid for.
- Concent Core is responsible for removing `DepositClaim` objects if the outcome of the use case indicates that they don't need to be paid (e.g. if additional verification confirms requestor's decision to reject the result).
- `Client` and `DepositAccount` objects are never removed.
    They stay in the database even if the operation fails and even after all the `DepositClaim`s that refer to them are removed.

**Comment by PaweuB** [comment]
Pivotal:
https://www.pivotaltracker.com/story/show/161280810
